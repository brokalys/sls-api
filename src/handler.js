import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
