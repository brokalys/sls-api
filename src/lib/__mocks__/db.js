export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
