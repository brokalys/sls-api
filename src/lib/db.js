import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
