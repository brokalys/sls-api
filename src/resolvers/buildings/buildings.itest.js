import { createTestClient } from 'apollo-server-testing';

import { server } from 'handler';
import db from 'lib/db';
import Bugsnag from 'lib/bugsnag';

jest.mock('lib/bugsnag');
jest.mock('lib/db');

describe('Query: buildings', () => {
  let run;

  beforeEach(() => {
    const utils = createTestClient(server);
    run = utils.query;
  });

  afterEach(jest.resetAllMocks);

  test('successfully retrieves building data', async () => {
    db.query.mockReturnValueOnce([
      {
        id: 1,
        bounds: [
          [
            { x: 56.992294, y: 24.136619 },
            { x: 56.976394, y: 23.99579 },
            { x: 56.992294, y: 24.136619 },
          ],
        ],
      },
      {
        id: 2,
        bounds: [
          [
            { x: 56.992294, y: 24.136619 },
            { x: 56.976394, y: 23.99579 },
            { x: 56.992294, y: 24.136619 },
          ],
        ],
      },
    ]);

    const response = await run({
      query: `
        query GetBuildings($bounds: String!) {
          buildings(bounds: $bounds) {
            id
            bounds
          }
        }
      `,
      variables: {
        bounds:
          '56.944756215513316 24.09404948877113, 56.9404253703127 24.09404948877113, 56.9404253703127 24.086952363717725, 56.944756215513316 24.086952363717725, 56.944756215513316 24.09404948877113',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves single building', async () => {
    db.query.mockReturnValueOnce([
      {
        id: 1,
        bounds: [
          [
            { x: 56.992294, y: 24.136619 },
            { x: 56.976394, y: 23.99579 },
            { x: 56.992294, y: 24.136619 },
          ],
        ],
      },
    ]);

    const response = await run({
      query: `
        query GetBuilding($id: Int!) {
          buildings(id: $id) {
            id
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response).toMatchSnapshot();
  });

  // @todo: passing headers to integration tests is currently not possible..
  // @see https://github.com/apollographql/apollo-server/issues/2277
  xtest('successfully returns nothing if there are no buildings in this bound', async () => {
    db.query.mockReturnValueOnce([]);

    const response = await run({
      query: `
        query GetBuildings($bounds: String!) {
          buildings(bounds: $bounds) {
            id
            properties {
              results {
                price
              }
            }
          }
        }
      `,
      variables: {
        bounds:
          '56.944756215513316 24.09404948877113, 56.9404253703127 24.09404948877113, 56.9404253703127 24.086952363717725, 56.944756215513316 24.086952363717725, 56.944756215513316 24.09404948877113',
      },
    });

    expect(response).toMatchSnapshot();
  });

  // @todo: passing headers to integration tests is currently not possible..
  // @see https://github.com/apollographql/apollo-server/issues/2277
  xtest('successfully retrieves building an property information', async () => {
    db.query.mockReturnValueOnce([
      {
        id: 1,
        bounds: [
          [
            { x: 56.992294, y: 24.136619 },
            { x: 56.976394, y: 23.99579 },
            { x: 56.992294, y: 24.136619 },
          ],
        ],
      },
      {
        id: 2,
        bounds: [
          [
            { x: 56.992294, y: 24.136619 },
            { x: 56.976394, y: 23.99579 },
            { x: 56.992294, y: 24.136619 },
          ],
        ],
      },
      {
        id: 3,
        bounds: [
          [
            { x: 56.992294, y: 24.136619 },
            { x: 56.976394, y: 23.99579 },
            { x: 56.992294, y: 24.136619 },
          ],
        ],
      },
    ]);
    db.query.mockReturnValueOnce([
      { price: 1, building_id: 1 },
      { price: 2, building_id: 1 },
      { price: 3, building_id: 2 },
    ]);

    const response = await run({
      query: `
        query GetBuildingsAndProperties($bounds: String!) {
          buildings(bounds: $bounds) {
            id
            properties {
              results {
                price
              }
            }
          }
        }
      `,
      variables: {
        bounds:
          '56.944756215513316 24.09404948877113, 56.9404253703127 24.09404948877113, 56.9404253703127 24.086952363717725, 56.944756215513316 24.086952363717725, 56.944756215513316 24.09404948877113',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('throws a validation exception if no bounds filter provided', async () => {
    const response = await run({
      query: `
        {
          buildings {
            id
          }
        }
      `,
    });

    expect(response).toMatchSnapshot();
    expect(db.query).not.toBeCalled();
  });

  test('throws a validation exception if bounds filter has too large location', async () => {
    const response = await run({
      query: `
        query GetBuildingsAndProperties($bounds: String!) {
          buildings(bounds: $bounds) {
            id
          }
        }
      `,
      variables: {
        bounds:
          '57.0510741522279 24.34369621296768, 56.86735048784755 24.34369621296768, 56.86735048784755 23.842917061051175, 57.0510741522279 23.842917061051175, 57.0510741522279 24.34369621296768',
      },
    });

    expect(response).toMatchSnapshot();
    expect(db.query).not.toBeCalled();
  });

  test('throws an authentication error if trying to retrieve results without authorizing', async () => {
    const response = await run({
      query: `
        query GetBuildingsAndProperties($bounds: String!) {
          buildings(bounds: $bounds) {
            id
            properties {
              results {
                price
              }
            }
          }
        }
      `,
      variables: {
        bounds:
          '56.944756215513316 24.09404948877113, 56.9404253703127 24.09404948877113, 56.9404253703127 24.086952363717725, 56.944756215513316 24.086952363717725, 56.944756215513316 24.09404948877113',
      },
    });

    expect(Bugsnag.notify).not.toBeCalled();
    expect(response.errors).toEqual([
      expect.objectContaining({
        extensions: {
          code: 'UNAUTHENTICATED',
        },
      }),
    ]);
  });
});
