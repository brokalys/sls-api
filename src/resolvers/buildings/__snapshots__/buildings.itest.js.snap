// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query: buildings successfully retrieves building an property information 1`] = `
Object {
  "data": Object {
    "buildings": Array [
      Object {
        "id": 1,
        "properties": Object {
          "results": Array [
            Object {
              "price": 1,
            },
            Object {
              "price": 2,
            },
          ],
        },
      },
      Object {
        "id": 2,
        "properties": Object {
          "results": Array [
            Object {
              "price": 3,
            },
          ],
        },
      },
      Object {
        "id": 3,
        "properties": Object {
          "results": Array [],
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: buildings successfully retrieves building data 1`] = `
Object {
  "data": Object {
    "buildings": Array [
      Object {
        "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619",
        "id": 1,
      },
      Object {
        "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619",
        "id": 2,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: buildings successfully returns nothing if there are no buildings in this bound 1`] = `
Object {
  "data": Object {
    "buildings": Array [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: buildings throws a validation exception if bounds filter has too large location 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "BAD_USER_INPUT",
        "details": Array [
          Object {
            "context": Object {
              "area": 0.0001,
              "key": "bounds",
              "label": "bounds",
              "value": "57.0510741522279 24.34369621296768, 56.86735048784755 24.34369621296768, 56.86735048784755 23.842917061051175, 57.0510741522279 23.842917061051175, 57.0510741522279 24.34369621296768",
            },
            "message": "\\"bounds\\" are must not exceed 0.0001",
            "path": Array [
              "bounds",
            ],
            "type": "string.maxArea",
          },
        ],
        "exception": Object {
          "details": Array [
            Object {
              "context": Object {
                "area": 0.0001,
                "key": "bounds",
                "label": "bounds",
                "value": "57.0510741522279 24.34369621296768, 56.86735048784755 24.34369621296768, 56.86735048784755 23.842917061051175, 57.0510741522279 23.842917061051175, 57.0510741522279 24.34369621296768",
              },
              "message": "\\"bounds\\" are must not exceed 0.0001",
              "path": Array [
                "bounds",
              ],
              "type": "string.maxArea",
            },
          ],
        },
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Input validation failed",
      "path": Array [
        "buildings",
      ],
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: buildings throws a validation exception if no bounds filter provided 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Field \\"buildings\\" argument \\"bounds\\" of type \\"String!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
